hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ExitUtil.java

```diff
@@ -159,138 +163,155 @@ public static void disableSystemHalt() {
    */
   public static void resetFirstExitException() {
-    firstExitException = null;
+    FIRST_EXIT_EXCEPTION.set(null);
   }
 
+  /**
+   * Reset the tracking of process termination. This is for use in unit tests
+   * where one test in the suite expects a halt but others do not.
+   */
   public static void resetFirstHaltException() {
-    firstHaltException = null;
+    FIRST_HALT_EXCEPTION.set(null);
   }
 
   /**
+   * Exits the JVM if exit is enabled, rethrow provided exception or any raised error otherwise.
    * Inner termination: either exit with the exception's exit code,
    * or, if system exits are disabled, rethrow the exception.
    * @param ee exit exception
+   * @throws ExitException if {@link System#exit(int)} is disabled and not suppressed by an Error
+   * @throws Error if {@link System#exit(int)} is disabled and one Error arise, suppressing
+   * anything else, even <code>ee</code>
    */
   public static void terminate(ExitException ee)
       throws ExitException {
     final int status = ee.getExitCode();
-    Error catched = null;
+    Error caught = null;
     if (status != 0) {
       try {
-        //exit indicates a problem, log it
+        // exit indicates a problem, log it
         String msg = ee.getMessage();
         LOG.debug("Exiting with status {}: {}",  status, msg, ee);
         LOG.info("Exiting with status {}: {}", status, msg);
       } catch (Error e) {
-        catched = e; // errors have higher priority than HaltException, it may be re-thrown. OOM and ThreadDeath are 2 examples of Errors to re-throw
+        // errors have higher priority than HaltException, it may be re-thrown.
+        // OOM and ThreadDeath are 2 examples of Errors to re-throw
+        caught = e;
       } catch (Throwable t) {
-        // all other kind of throwables are supressed
+        // all other kind of throwables are suppressed
         ee.addSuppressed(t);
       }
     }
     if (systemExitDisabled) {
       try {
         LOG.error("Terminate called", ee);
       } catch (Error e) {
-        if (catched == null) {
-          catched = e; // errors will be re-thrown
+        // errors have higher priority again, if it's a 2nd error, the 1st one suprpesses it
+        if (caught == null) {
+          caught = e;
         } else {
-          catched.addSuppressed(e); // 1st raised error has priority and will be re-thrown, so the 1st error supresses the 2nd
+          caught.addSuppressed(e);
         }
       } catch (Throwable t) {
-        ee.addSuppressed(t); // all other kind of throwables are supressed
-      }
-      synchronized (ExitUtil.class) {
-        if (!terminateCalled()) {
-          firstExitException = ee;
-        }
+        // all other kind of throwables are suppressed
+        ee.addSuppressed(t);
       }
-      if (catched != null) {
-        catched.addSuppressed(ee);
-        throw catched;
+      FIRST_EXIT_EXCEPTION.compareAndSet(null, ee);
+      if (caught != null) {
```

Comment:
do you think we need a safety check to see if ee != caught before addSuppressed()? we've had problems with abfs in the past, where a cached exception was being rethrown in close()

Reply:
Hi, sorry to have taken so long before an update.

I don't think we need such a check because the only things able to throw something would be either the JVM (which never reuse already thrown things) or the LOG object (which should never do that too). In my understanding of exception and any throwable in general is to never throw 2 times the same instance. If such a rule is respected, a suppressed can never a suppressor of itself through any suppressor/suppressed/rootCause chain of throwables. If for some reason the addSuppressed fails because of this, this more means there is something broken implemented elsewhere and this "elsewhere" should be fixed. If some piece of code reuse some already thrown throwables, even without failing, the suppress stuff could lead to memory leak (adding again and again new throwables to a singleton one kept and reuse somewhere).

In fact, if we really want to be safe, then we can forget the suppression stuff : this would keep code simple, no broken suppression error, no leak, at the cost of a slight loss of debug info when dumping throwable.

maybe we should rather go for this safe solution.

what do you think ?

Reply:
Finally I've done what you suggested since its efficient enough, robust enough, et does not clobber that much the code.

Reply:
> understanding of exception and any throwable in general is to never throw 2 times the same instance.

some classes cache a raised exception and raised later. see https://issues.apache.org/jira/browse/HADOOP-16785

(I do agree with the "elsewhere" claim, just want to be resilient to the mistake)